### ISSUE TITLE
postMessage with large data buffers can crash chrome

### ISSUE THREAD

**Issue by titlestad** [photopea/photopea#652]
When sending large data buffers, we may end up in a deadlock situation. This is because of the single thread behavior of browsers on some functionality like postMessage...
To get around this, there needs to be chunking of the data.
This in fact has a great benefit as a bonus - we can make progress bars!
I suggest a chunking scheme for big data files (like a large PSD):

```
var uniqueId = 1;

// Settings
var chunkSize = 4096; // 4k chunks

// Fragment
var messagePrototype = {
  command: 'filetransfer'
};

// This will keep all of our ongoing transfers
var dataQueue = {};

// This is your message handler connected to onmessage
function onMessageHandler( msg )
{
  if( msg.command == 'filetransfer' )
  {
    // Ignore
    if( !dataQueue[ msg.uniqueId ] ) return;
    if( msg.chunkCount - 1 == msg.chunk )
    {
      // Done!
      delete dataQueue[ msg.uniqueId ];
    }
    else
    {
      // Continue sending!
      sendChunkedMessage( null, transferId );
    }
  }
}

// Do the actual sending
sendChunkedMessage( myData );

// Send function
function sendChunkedMessage( data, transferId )
{
  // First chunk has no transferId
  if( !transferId )
  	transferId = uniqueId; // Next available id
  
  // No dataqueue here? It's the first chunk
  if( !dataQueue[ transferId ] )
  {
    // Prepare and send first chunk
    var msg = messagePrototype;
    msg.uniqueId = transferId;
    msg.chunkCount = Math.ceil( data.length / chunkSize );
    if( msg.chunkCount == 0 ) msg.chunkCount = 1;
    msg.chunk = 0;
    msg.length = data.length < chunkSize ? data.length : chunkSize;
    msg.data = data.substr( 0, msg.length );
    
    // Increment unique id
    uniqueId++;
    
    // We wrap around on 1024 (no need to increment indefinately)
    if( uniqueId > 1024 ) uniqueId = 0;
    
    // Store in queue
    dataQueue[ uniqueId ] = {
      msg: msg,
      wholeData: data
    };
    
    parent.postMessage( msg, '*' );
    return;
  }
   
  // Get the dataqueue
  var info = dataQueue[ msg.uniqueId ];
  
  // No dataqueue? Something went wrong!
  if( !info ) return;
 
  // Send next chunk
  info.msg.chunk++;
  var chunkOffset = chunkSize * info.msg.chunk;
  var leftOverDataSize = info.wholeData.length - chunkOffset;
  info.msg.data = info.wholeData.substr( chunkOffset, leftOverDataSize );
  info.msg.length = leftOverDataSize;
  parent.postMessage( info.msg, '*' );
}
```

**Comment by photopea** [comment]
Please, don't throw any code at us before explaining what it does and what it is good for! :)

How large the ArrayBuffer was in your case, when it crashed the browser? Are you sure the problem was in the buffer size? If yes, we should report this to the authors of your browser. I can not see what deadlock are you talking about.

**Comment by titlestad** [comment]
I don't think you will get anything back from the browser developers. This is a typical nofix. Sending 20 megs over both websockets or postMessage will require chunking. That's just the way of life, as the browser is trying to transfer the data at optimum speed.

My buffer was probably 20-30 megabytes. Or more. Photopea doesn't have debug output on that.

My proposal fixes this with dividing the transfer in smaller chunks. Which means we can make progress bars showing how much data has been transferred from total.

**Comment by titlestad** [comment]
Another thing that could lead to a crash is if there's a loop preparing the data for postMessage() which takes too long.

**Comment by photopea** [comment]
The postMessage() should not send anything at all. When it is called, the data is already located at the same computer, so the browser should just allow the "receiver of data" to access them. There is no need for a progress bar, because it should happen instantly (in less than 1ms, even for 4 GB of data).

The delay between you call postMessage() and receive "done" is the time Photopea processes the file, not sending it.

Does it crash every time you send a file larger than 20MB?

**Comment by titlestad** [comment]
I will have to debug further. Please keep this issue open until I can come back with more detailed data.
If it isn't postMessage, it may be the code that prepares the data for transfer.

**Comment by jimmywarting** [comment]
fyi, i have seen in some [example](https://photopea-api-demo.glitch.me/demo2.html) of where an arrayBuffer is being transfered with 
```js
function loaded(e) {
  var ab = e.target.response;  // ArrayBuffer
  wnd.postMessage(ab, "*");
}
```
this is a tiny bit inefficient, as it has to "clone" the arraybuffer memory so it takes up 2x as much ram.
postMessage has a transferable option too. meaning that it will be transfered instead of copied. and so that the main thread are no longer able to read the arrayBuffer (the size will become 0 on the main thread)

all you got to do is:   `wnd.postMessage(ab, "*", [ab])`
this might help with some memory issues...

---

if we really don't want to allocate any memory, then it would be better to share a blob / files, as they can just be pointers to somewhere backed up by the filesystem and dose not need to be fully read.
